rsl-learning2022-Risheel/TimerAndGame/app/src/main/java/com/example/timerandgame/MainActivity.java

```diff
@@ -0,0 +1,175 @@
+    private TextView time;
+    public static LinearLayout linearLayout;
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+        start=findViewById(R.id.start);
+        stop=findViewById(R.id.stop);
+        time=findViewById(R.id.time_text);
+        linearLayout=(LinearLayout) findViewById(R.id.main_layout);
+        ViewModelProvider viewModelProvider = new ViewModelProvider(
+                this, (ViewModelProvider.Factory) ViewModelProvider.AndroidViewModelFactory.getInstance(getApplication()));
+        viewModel = viewModelProvider.get(CountryListViewModel.class);
+        timerViewModel=viewModelProvider.get(TimerViewModel.class);
+        timerViewModel.getTimeData().observe(this,this::setTime);
+        start.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                timerViewModel.startTimer();
+            }
+        });
+        stop.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                timerViewModel.stopTimer();
+            }
+        });
+        setUpLiveData();
+    }
+    public void setTime(long millis){
+
+        String timeString = DateUtils.formatElapsedTime(millis/1000);
+        String millisString=String.format(Locale.ROOT,"%03d",millis%1000);
+
+        SpannableString formatString=new SpannableString(timeString+millisString);
+        formatString.setSpan(new RelativeSizeSpan(0.5f),timeString.length(),formatString.length(),0);
+        Log.d("OKOKOK", "setTime: "+formatString);
+        if(time!=null)
+        {
+            time.setText(formatString);
+        }
+    }
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        if (failureDialog != null) {
+            failureDialog.dismiss();
+        }
+        if (loadingDialog != null) {
+            loadingDialog.dismiss();
+        }
+    }
+
+    private void setUpLiveData() {
+        viewModel.getCountryLiveData().observe(this, new Observer<List<CountryModel>>() {
+            @Override
+            public void onChanged(List<CountryModel> countryModels) {
+                handleCountryList(countryModels);
+            }
+        });
+        viewModel.getRequestStatusLiveData().observe(this, new Observer<RequestStatus>() {
+            @Override
+            public void onChanged(RequestStatus requestStatus) {
+                handleRequestStatus(requestStatus);
+            }
+        });
+    }
+
+    private void handleCountryList(List<CountryModel> countryModels) {
+        CountryListAdapter adapter = new CountryListAdapter(countryModels,this);
+        RecyclerView countryRecyclerView = findViewById(R.id.gamesRecyclerView);
+        countryRecyclerView.setLayoutManager(new LinearLayoutManager(this));
+        countryRecyclerView.setAdapter(adapter);
```

Comment:
Avoid adapter initialization code in this callback. The network livedata could return data multiple times e.g. auto refresh, retry after fail, pull to refresh.